pkg/recordbot/recorder.go

```diff
@@ -0,0 +1,159 @@
+package recordbot
+
+import (
+	"errors"
+	"log"
+	"sync"
+
+	"github.com/pion/webrtc/v3"
+	"github.com/pion/webrtc/v3/pkg/media"
+	"github.com/pion/webrtc/v3/pkg/media/h264writer"
+	"github.com/pion/webrtc/v3/pkg/media/ivfwriter"
+	"github.com/pion/webrtc/v3/pkg/media/oggwriter"
+)
+
+var ErrUnsupportedCodec = errors.New("unsupported codec")
+
+type Recorder interface {
+	// Reads RTP packets in TrackRemote and save them to local file.
+	// No need to call this in a goroutine as the function already calls one.
+	Start(track *webrtc.TrackRemote)
+
+	// Stops the recording and does cleanups such as closing the file and channel.
+	Stop()
+}
+
+type recorder struct {
+	// Save file name to be passed into hooks
+	filename string
+
+	// Each recorder uses different writer depending if it's for video / audio
+	writer media.Writer
+
+	// The `done` channel is an internal channel for tracking the start to finish of the recording
+	done chan bool
+
+	// WaitGroup is used to ensure everything is cleaned up in the goroutines before closing the recorder
+	wg sync.WaitGroup
+
+	// Hooks provide access to customising aspects of the recorder, such as uploading the file after recording
+	hooks RecorderHooks
+}
+
+type RecorderHooks struct {
+	// The recorder will pass the filename that it was instantiated with, to the hook
+	UploadFile func(filename string) error
+}
+
+func NewRecorder(filename string, codec webrtc.RTPCodecParameters, hooks RecorderHooks) (Recorder, error) {
+	// Instantiates the appropriate media writer to save RTP packets
+	writer, err := createMediaWriter(filename, codec.MimeType)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make channel for tracking recorder state
+	done := make(chan bool, 1)
+
+	return &recorder{filename, writer, done, sync.WaitGroup{}, hooks}, nil
+}
+
+func createMediaWriter(fileName string, mimeType string) (media.Writer, error) {
+	if mimeType == webrtc.MimeTypeVP8 || mimeType == webrtc.MimeTypeVP9 {
+		return ivfwriter.New(fileName)
+	}
+	if mimeType == webrtc.MimeTypeH264 {
+		return h264writer.New(fileName)
+	}
+	if mimeType == webrtc.MimeTypeG722 ||
+		mimeType == webrtc.MimeTypePCMA ||
+		mimeType == webrtc.MimeTypePCMU ||
+		mimeType == webrtc.MimeTypeOpus {
+		// The values for oggwriter can be found at
+		// https://github.com/pion/webrtc/blob/master/examples/save-to-disk/main.go#L53
+		return oggwriter.New(fileName, 48000, 0)
+	}
+	return nil, ErrUnsupportedCodec
+}
+
+func (r *recorder) Start(track *webrtc.TrackRemote) {
+	r.wg.Add(1)
+	go r.record(track)
+}
+
+func (r *recorder) record(track *webrtc.TrackRemote) {
```

Comment:
I think this loop could be written a bit differently to be easier to read.

You can keep these struct members:
* closedChan chan struct{}
* done chan struct{}

`done` has a single purpose of notifying that a user has requested stop. In this case it's more idiomatic to close it since that'll notify all pending readers.

```
var err error
defer func() {
    // handle error
    if err != nil {
    }
    // move cleanup routine in here
    r.cleanup()
    if r.hooks.UploadFile != nil ...
    
    // signal Stop that this is done
    close(r.closeChan)
}()
for {
    select {
    case <-done:
        return
    default:
        // Read RTP stream
        var packet rtp.Packet
	packet, _, err = track.ReadRTP()
	if err != nil {
		return
	}

	// Write to file
	err = writer.WriteRTP(packet)
	if err != nil {
		return
	}
    }
}
```

Reply:
Hi @davidzhao, I didn't implement the struct members as I was unsure why this is the recommended practice. Perhaps you can clarify this ?
- Why is using an empty struct better than a boolean ?
- Why do we need both `closedChan` and `done` ? I would assume `done` is sufficient just for tracking the `stop` signal.

Reply:
struct{} is preferred as it only indicates the type, but consumes 0 bytes. The main point I'm trying to make is closing the channel instead of sending a value to it. Closing ensures two things:
1. it'll not block, you can always close a channel. writing to a channel would block if it's unbuffered
2. close would notify all readers that they can no longer expect to get values from it

`closeChan` is used to allow `Stop()` to block, until post processing is finished. For example, if you were to Stop the recorder, then immediately read the file from S3, it may error out if Stop returned before the write operation is complete

Reply:
I think I understand. This means I can use either buffered or unbuffered channels for `done` and I shouldn’t run into deadlock right ? But for `closedChan`, it will have to be an unbuffered channel and I’ll have something like `<- closedChan` in `Stop()` to block the cleanup process until it’s done.